Ref updates for 7.2

BillWagner opened the pull request:
Fixes #3241 

Explain the concepts behind `in` parameters, `ref readonly` returns, `readonly struct` types and `ref struct` types.

BillWagner committed:
first draft ready

BillWagner committed:
add samples

Ready for a test build.

BillWagner committed:
give it a proofread

BillWagner committed:
Fix the build errors.

BillWagner commented:
I checked the build and the links, this is ready to review.

/cc @VSadov

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+
+The compiler enforces that the caller cannot modify the reference. Attempts
+to assign to the value directly generate a compile-time error. Calling member
+methods creates a defensive copy. 
```
define defensive copy

BillWagner replied:
I updated the paragraph to define it.

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+can call any instance method that uses pass-by-value semantics. In
+those instances, a copy of the `in` parameter is created. Because the compiler
+can create a temporary for any `in` parameter, you can also specify default
+values for any `in` parameter. The follow code uses that to specify the origin
```
define what origin is here, just to make it clear -- i.e, origin (point 0,0).

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+It's likely that the library using `Point3D` would often use the origin
+throughout the code. Every instance creates a new object on the stack. It may
+be advantageous to create a constant and return it by reference. But, if you
+are returning a reference to internal storage, you may want to enforce that
```
are returning --> return

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+to assign to the value directly generate a compile-time error. Calling member
+methods creates a defensive copy. 
+
+Creating a copy of a ref readonly return is easy: Just assign it to a variable
```
I'd move this to follow the example. It's best to have the example -- which is a good one -- follow immediately to provide a use case for ref readonly and make it less abstract.

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+
+[!code-csharp[OriginReference](../../samples/csharp/reference-semantics/Point3D.cs#OriginReference "Creating a readonly Origin reference")]
+
+When you assign a variable to a `ref return`, you can specify either a `ref readonly`
```
is `ref return` `ref readonly return`?

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+
+[!code-csharp[AssignRefReadonly](../../samples/csharp/reference-semantics/Program.cs#AssignRefReadonly "Assigning a ref readonly")]
+
+The first assignment in the preceding code makes a copy of the `Origin` constant, and assigns
```
no comma after constant

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+The first assignment in the preceding code makes a copy of the `Origin` constant, and assigns
+that copy. The second assigns a reference. Notice that the `readonly` modifier
+must be part of the declaration of the variable. The reference to which it refers
+cannot be modified. Attempts to do so results in a compile-time error.
```
results --> result

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+## `readonly struct` type
+
+Applying `ref readonly` to high-traffic uses of a struct may be sufficient.
+Other times, you may want to create an immutable struct. Then, you can
```
no comma after then

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+that take place when you access methods of a struct used as an `in` parameter.
+
+You can do that by creating a `readonly struct` type. You can add the `readonly`
+modifier to a struct declaration. The language enforces that all members of
```
compiler enforces?

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+
+You can do that by creating a `readonly struct` type. You can add the `readonly`
+modifier to a struct declaration. The language enforces that all members of
+the struct are `readonly`: the `struct` must be immutable.
```
semicolon instead of colon.

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+`readonly struct`: The `this` reference is always an `in` parameter, instead
+of a copy of the receiver, passed by value to the member method. This optimization
+saves more copying when you use a `readonly struct`. The `Point3D` would be a great
+candidate for this change. The following code shows an update `ReadonlyPoint3D` structure:
```
update --> updated

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+returning the object.
+
+Finally, the compiler generates more efficient code when you call members of a
+`readonly struct`: The `this` reference is always an `in` parameter, instead
```
Hard to parse. Perhaps "The `this` reference, instead of a copy of the receiver, is always an `in` parameter passed by value to the member method.

BillWagner replied:
Also, it should be "passed by reference" not "passed by value".

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+Finally, the compiler generates more efficient code when you call members of a
+`readonly struct`: The `this` reference is always an `in` parameter, instead
+of a copy of the receiver, passed by value to the member method. This optimization
+saves more copying when you use a `readonly struct`. The `Point3D` would be a great
```
nit: would be --> is

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+must be stack allocated. In other words, these types can never be created on the
+heap as a member of another class. The primary motivation for this feature
+was `Span<T>` and related structures, but you can create `ref struct` types
+when you have similar requirements. In this article, you see examples
```
The requirements aren't clear. Can you mention a use case here?

BillWagner replied:
I re-wrote that section. Likely, using `Span<T>` and related are the most common use case. I tried to make that point, and point to other uses.

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+types declared as `ref struct` include `ReadOnlySpan<T>`, `Memory<T>`, and
+`ReadOnlyMemory<T>`. 
+
+The goal of keeping a `ref struct` type as a stack allocated variable
```
stack allocated --> stack-allocated

rpetrusha commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+algorithms where memory allocations can be critical to achieving the
+necessary performance. You may find that you don't often use these features in
+the code you write. However, these enhancements have been adopted in many locations
+in the .NET framework. As more and more APIs make use of these features, you'll
```
framework --> Framework

rpetrusha reviewed (changes requested):
This looks good. Left a number of comments, @BillWagner.

BillWagner commented on docs/csharp/ref-locals-and-returns.md:
```diff
@@ -0,0 +1,205 @@
+When you add the `in` modify to pass an argument by reference, you declare
+your design intent is to pass arguments by reference to
+avoid unnecessary copying. You do not intend to modify the object used
+as that argument. The following code shows an example of a method
```
Good point. I added more context in the paragraph above.

BillWagner committed:
respoond to feedback

BillWagner committed:
rename file

"reference-semantics-with-value-types.md" better reflects the content.

BillWagner commented:
@rpetrusha I updated everything based on your comments. Let me know if this is ready.

I also changed the file name in a separate commit so it should be easier to review.

jcouv commented on docs/csharp/reference-semantics-with-value-types.md:
````diff
@@ -0,0 +1,213 @@
+  </PropertyGroup>
+```
+
+You can use either "7.2" or "latest" for the value.
````
Given that there are multiple ways of updating a project (UI settings, editing csproj, code fix), consider factoring this section to a dedicated page, and referenced from here.
Then you can just have one line: "In order to use those features, you have to configure your project to use C# 7.2 or later in your project, by setting the language version to "7.2" or "latest" per _these instructions_".

BillWagner replied:
I'd agree. 

I want to take a bit more time to do that, so I made an issue that can get addressed in this or the next sprint: #3685

rpetrusha reviewed (approved):
LGTM, @BillWagner. This is good to merge when you're ready, after the build completes.

BillWagner committed:
fix a build error
